File: sdks/python-mcp/python/solvapay_mcp/server/narrate.py
from __future__ import annotations

from collections.abc import Callable, Mapping

from solvapay_mcp.core import call

NarratorOutput = dict[str, object]


def _as_mapping(value: object) -> dict[str, object] | None:
    if isinstance(value, Mapping):
        return {str(k): v for k, v in value.items()}
    return None


def _as_narrator_output(value: object) -> NarratorOutput:
    mapping = _as_mapping(value)
    if mapping is None:
        raise TypeError("mcpNarrate returned a non-object narrator envelope")
    text = mapping.get("text")
    result: NarratorOutput = {"text": text if isinstance(text, str) else ""}
    links = mapping.get("links")
    if isinstance(links, list) and links:
        result["links"] = links
    return result


def _narrate(tool: str, data: Mapping[str, object]) -> NarratorOutput:
    return _as_narrator_output(call("mcpNarrate", {"tool": tool, "payload": dict(data)}))


def narrate_manage_account(data: Mapping[str, object]) -> NarratorOutput:
    return _narrate("manage_account", data)


def narrate_upgrade(data: Mapping[str, object]) -> NarratorOutput:
    return _narrate("upgrade", data)


def narrate_topup(data: Mapping[str, object]) -> NarratorOutput:
    return _narrate("topup", data)


def narrate_activate_plan(data: Mapping[str, object]) -> NarratorOutput:
    return _narrate("activate_plan", data)


NARRATORS: dict[str, Callable[[Mapping[str, object]], NarratorOutput]] = {
    "upgrade": narrate_upgrade,
    "manage_account": narrate_manage_account,
    "topup": narrate_topup,
    "activate_plan": narrate_activate_plan,
}


def ui_placeholder(tool: str, data: Mapping[str, object]) -> str:
    result = _as_narrator_output(
        call("mcpNarrate", {"tool": tool, "payload": dict(data), "kind": "placeholder"})
    )
    text = result.get("text")
    return text if isinstance(text, str) else ""


def balance_summary(customer: Mapping[str, object] | None) -> str | None:
    result = _as_narrator_output(
        call(
            "mcpNarrate",
            {
                "tool": "manage_account",
                "payload": dict(customer) if customer is not None else {},
                "kind": "balanceSummary",
            },
        )
    )
    text = result.get("text")
    return text if isinstance(text, str) and text else None

File: sdks/typescript/mcp-core/src/narrate.ts
/**
 * Intent-tool narrators — thin wrappers over the Rust `mcpNarrate` op.
 * Markdown is user-visible in agent transcripts; do not reimplement it here.
 */

import { callMcpSyncOp } from './native-mcp'
import type { BootstrapPayload } from './types'

export interface NarratorOutput {
  text: string
  links?: Array<{ uri: string; name: string }>
}

export type IntentTool = 'upgrade' | 'manage_account' | 'topup' | 'activate_plan'

type NarrateEnvelope = {
  text?: unknown
  links?: unknown
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

function asNarratorOutput(value: unknown): NarratorOutput {
  if (!isRecord(value)) {
    throw new Error('mcpNarrate returned a non-object narrator envelope')
  }
  const text = typeof value.text === 'string' ? value.text : ''
  const links = Array.isArray(value.links)
    ? value.links.flatMap(link => {
        if (!isRecord(link)) return []
        const uri = typeof link.uri === 'string' ? link.uri : ''
        const name = typeof link.name === 'string' ? link.name : ''
        return uri && name ? [{ uri, name }] : []
      })
    : undefined
  return links && links.length > 0 ? { text, links } : { text }
}

function narrate(tool: IntentTool, data: BootstrapPayload): NarratorOutput {
  return asNarratorOutput(callMcpSyncOp<NarrateEnvelope>('mcpNarrate', { tool, payload: data }))
}

export function narrateManageAccount(data: BootstrapPayload): NarratorOutput {
  return narrate('manage_account', data)
}

export function narrateUpgrade(data: BootstrapPayload): NarratorOutput {
  return narrate('upgrade', data)
}

export function narrateTopup(data: BootstrapPayload): NarratorOutput {
  return narrate('topup', data)
}

export function narrateActivatePlan(data: BootstrapPayload): NarratorOutput {
  return narrate('activate_plan', data)
}

export const NARRATORS: Record<IntentTool, (data: BootstrapPayload) => NarratorOutput> = {
  upgrade: narrateUpgrade,
  manage_account: narrateManageAccount,
  topup: narrateTopup,
  activate_plan: narrateActivatePlan,
}

export function uiPlaceholder(tool: IntentTool, data: BootstrapPayload): string {
  const result = callMcpSyncOp<NarrateEnvelope>('mcpNarrate', {
    tool,
    payload: data,
    kind: 'placeholder',
  })
  return typeof result.text === 'string' ? result.text : ''
}

export function balanceSummary(customer: { balance?: unknown } | null | undefined): string | null {
  const result = callMcpSyncOp<NarrateEnvelope>('mcpNarrate', {
    tool: 'manage_account',
    payload: customer ?? {},
    kind: 'balanceSummary',
  })
  return typeof result.text === 'string' ? result.text : null
}
